pydantic/dataclasses.py

```diff
@@ -0,0 +1,43 @@
+from dataclasses import dataclass as dc
+from .main import validate_model, create_model
+
+
+def pydantic_model(cls):
+    fields = {name: (field.type, field.default)
+              for name, field
+              in cls.__dataclass_fields__.items()}
+    model = create_model('', **fields)
+    return model
+
+
+def dataclass(_cls=None, *, init=True, repr=True, eq=True,
+              order=False, unsafe_hash=False, frozen=False):
+    def wrap(cls):
+        post_init_orig = getattr(cls, '__post_init__', lambda self: None)
+        cls.__post_init__ = post_init_orig
+        cls = dc(init=init, repr=repr, eq=eq, order=order,
+                 unsafe_hash=unsafe_hash, frozen=frozen)(cls)
+        model = pydantic_model(cls)
+
+        def set_attr(self, name, value):
+            super(cls, self).__setattr__(name, value)
+            if getattr(self, '__initialized', False):
+                validate_model(model, self.__dict__)
+
+        def post_init(self):
+            post_init_orig(self)
+            validate_model(model, self.__dict__)
```

Comment:
Individual field validation will require a reference to the model instance to be kept, something like

```py
self.__model__ = validate_model(model, self.__dict__)
```

Reply:
the model is the same for all instances. The validate_model creates some copies, isn't it wasteful to store it to each instance?

Reply:
I guess it's more memory, but it avoids lots of extra processing in `set_attr`.

It could also lead to some unexpected behaviour, for example: let's say you implement your own custom type `Foobar` and use it on a dataclass:

```py
class Foobar:
    @classmethod
    def get_validators(cls):
        yield cls.validate

    @classmethod
    def validate(cls, value):
        # do some validation
        ...

@dataclass
def MyDataclass:
    foobar: Foobar
    other: str

```

Now you expect `validate` to be called once on 
```py
mydataclass = MyDataclass(foobar=xyz, other='abc')
```
but you don't expect it to be called on 
```py
mydataclass.other = 'whatever'
```

With the current implementation the whole model will be validated every time a single field is set.

Reply:
I think that the setattr is not that important in most dataclass uses. Most people will use immutable dataclasses. Can't we just hard set frozen=True?

Reply:
hard settings `frozen=True` with an exception when people try settings `frozen=False` is fine with me. `NotImplementedError` or is that confusing?

The value coercion point above however is more important and more tricky to solve.